avogadro/tdxcontroller.cpp

```diff
@@ -0,0 +1,592 @@
+  extents.min.y = m_pGLRenderer->m_orthographicFrustum[2];
+  extents.min.z = -m_pGLRenderer->m_orthographicFrustum[5];
+  extents.max.x = m_pGLRenderer->m_orthographicFrustum[1];
+  extents.max.y = m_pGLRenderer->m_orthographicFrustum[3];
+  extents.max.z = -m_pGLRenderer->m_orthographicFrustum[4];
+  return 0;
+}
+
+long Avogadro::TDxController::GetViewFOV(double &fov) const
+{
+  return navlib::make_result_code(navlib::navlib_errc::no_data_available);
+}
+
+long Avogadro::TDxController::GetViewFrustum(navlib::frustum_t &frustum) const
+{
+  if (m_pGLRenderer == nullptr)
+    return navlib::make_result_code(navlib::navlib_errc::no_data_available);
+
+  frustum.left = m_pGLRenderer->m_perspectiveFrustum[0];
+  frustum.right = m_pGLRenderer->m_perspectiveFrustum[1];
+  frustum.bottom = m_pGLRenderer->m_perspectiveFrustum[2];
+  frustum.top = m_pGLRenderer->m_perspectiveFrustum[3];
+  frustum.nearVal = m_pGLRenderer->m_perspectiveFrustum[4];
+  frustum.farVal = m_pGLRenderer->m_perspectiveFrustum[5];
+  return 0;
+}
+ 
+long Avogadro::TDxController::GetIsViewPerspective(
+  navlib::bool_t &perspective) const
+{
+  if (m_pGLRenderer == nullptr)
+    return navlib::make_result_code(navlib::navlib_errc::no_data_available);
+
+  perspective = (m_pGLRenderer->camera().projectionType() ==
+                 Avogadro::Rendering::Projection::Perspective);
+  return 0;
+}
+
+
+TDx::SpaceMouse::CCategory Avogadro::TDxController::getCategory(
+  const std::string &pathCode,
+  const std::shared_ptr<ActionTreeNode> &pNode)
+{
+
+  TDx::SpaceMouse::CCategory result(pNode->m_nodeName, pNode->m_nodeName);
+  std::string nextPathCode(pathCode);
+
+  for (uint32_t i = 0u; i < pNode->m_children.size(); i++) {
+    nextPathCode.append(std::to_string(i));
+    nextPathCode.push_back('.');
+    result.push_back(getCategory(nextPathCode, pNode->m_children[i]));
+    nextPathCode = pathCode;
+  }
+
+  for (uint32_t i = 0u; i < pNode->m_actions.size(); i++) {
+    
+	if (pNode->m_actions[i]->iconText().isEmpty())
+      continue;
+
+	std::string commandId(nextPathCode + std::to_string(i));
+
+    result.push_back(TDx::SpaceMouse::CCommand(
+        commandId,
+		pNode->m_actions[i]->iconText().toStdString(),
+		pNode->m_actions[i]->toolTip().toStdString()));
+        
+#ifdef WIN32
+
+	const QIcon iconImg = pNode->m_actions[i]->icon();
+    const QImage qimage = iconImg.pixmap(QSize(256, 256)).toImage();
+    QByteArray qbyteArray;
+    QBuffer qbuffer(&qbyteArray);
+    qimage.save(&qbuffer, "PNG");
+
+    TDx::CImage icon = TDx::CImage::FromData(qbyteArray.toStdString(), 0, commandId.c_str());
+
+    m_utilityIcons.push_back(icon);
+#endif
+  }
+  return result;
+}
+
+long Avogadro::TDxController::GetModelExtents(navlib::box_t &extents) const
+{
+  if (m_pGLRenderer == nullptr)
+    return navlib::make_result_code(navlib::navlib_errc::no_data_available);
+
+  std::vector<bool> flags;
+
+  m_pGLRenderer->scene().getBoundingBox(extents.min.x, 
```

Comment:
See, I think you can get this from the molecule here and avoid having to save all those spheres in the render pipeline.

Just implement `core/molecule.*:getBoundingBox()` instead.

Reply:
Done as suggested.